### PR TITLE
executor: Add sendRate for copIterator in non-keepOrder

### PR DESCRIPTION
**Pull request by Yisaer**
Problem Summary:

### What is changed and how it works?
Add sendRate function for copIterator when the keep Order is false.  

This requests change the copIterator's `Next` function. The copResponse would be stored into each task's `respCh` instead of `copIterator`'s own `respCh` so that the sendRate can work in both keepOrder or non-KeepOrder cases by `getToken` before sender putting task and `putToken` after one task finished.

To receive task from task's respCh with non-keepOrder, I introduce `copResponseCollector` to collect response from all the tasks.   `copResponseCollector` will collect the response into its `respCh` and `copIterator` can receive response from collector's respCh in its own `Next` function like before.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

-

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -491,12 +493,7 @@ const minLogCopTaskTime = 300 * time.Millisecond
 func (worker *copIteratorWorker) run(ctx context.Context) {
 	defer worker.wg.Done()
 	for task := range worker.taskCh {
-		respCh := worker.respChan
-		if respCh == nil {
-			respCh = task.respChan
-		}
-
-		worker.handleTask(ctx, task, respCh)
+		worker.handleTask(ctx, task, task.respChan)
```
Comment (wshwsh12) [inline store/tikv/coprocessor.go:496]: Why not PutToken()  here so that we don't need copResponseCollector?

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -86,10 +86,14 @@ func (c *CopClient) Send(ctx context.Context, req *kv.Request, vars *kv.Variable
 		// Make sure that there is at least one worker.
 		it.concurrency = 1
 	}
-	if it.req.KeepOrder {
-		it.sendRate = newRateLimit(2 * it.concurrency)
-	} else {
-		it.respChan = make(chan *copResponse, it.concurrency)
+	it.sendRate = newRateLimit(it.concurrency)
```
Comment (fzhedu) [inline store/tikv/coprocessor.go:89]: it change the rate `newRateLimit(2 * it.concurrency)`

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -1176,3 +1168,50 @@ func (it copErrorResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
...
+
+// copResponseCollector is used to collect the resp from task respCh into its own resp
+type copResponseCollector struct {
+	respChan chan *copResponse
+	tasks    []*copTask
+	sendRate *rateLimit
+	finishCh <-chan struct{}
+}
+
+func (c *copResponseCollector) run() {
```
Comment (fzhedu) [inline store/tikv/coprocessor.go:1180]: it should use with.recovery to record the heap if the goroutine exit by accident.

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (Yisaer) [reply store/tikv/coprocessor.go:496]: good catch, pull https://github.com/pingcap/tidb/pull/19244 request instead.

**Pull request closed without merging**